soc/riscv/esp32c3/soc.c

```diff
@@ -0,0 +1,189 @@
+#include <soc.h>
+
+#define ESP32C3_INTC_DEFAULT_PRIO 15
+
+extern void z_cstart(void);
+extern void esprv_intc_int_set_threshold(int priority_threshold);
+
+/*
+ * This is written in C rather than assembly since, during the port bring up,
+ * Zephyr is being booted by the Espressif bootloader.  With it, the C stack
+ * is already set up.
+ */
+void __attribute__((section(".iram1"))) __start(void)
+{
+	volatile uint32_t *wdt_rtc_protect = (uint32_t *)RTC_CNTL_WDTWPROTECT_REG;
+	volatile uint32_t *wdt_rtc_reg = (uint32_t *)RTC_CNTL_WDTCONFIG0_REG;
+	extern uint32_t _bss_start;
+	extern uint32_t _bss_end;
+
+	/* Configure the global pointer register
+     * (This should be the first thing startup does, as any other piece of code could be
+     * relaxed by the linker to access something relative to __global_pointer$)
+	 */
+	__asm__ __volatile__(".option push\n"
+			     ".option norelax\n"
+			     "la gp, __global_pointer$\n"
+			     ".option pop");
+
+	__asm__ __volatile__("la t0, _esp32c3_vector_table \n"
+				 "csrw mtvec, t0 \n");
+
+	/* Disable normal interrupts. */
+	csr_read_clear(mstatus, MSTATUS_MIE);
+
+	/* Zero out BSS.  Clobber _bss_start to avoid memset() elision. */
+	(void)memset(&_bss_start, 0, (&_bss_end - &_bss_start) * sizeof(_bss_start));
+	__asm__ __volatile__("" : : "g"(&_bss_start) : "memory");
+
+#if !CONFIG_BOOTLOADER_ESP_IDF
+	/* The watchdog timer is enabled in the 1st stage (ROM) bootloader.
+	 * We're done booting, so disable it.
+	 * If 2nd stage bootloader from IDF is enabled, then that will take
+	 * care of this.
+	 */
+	volatile uint32_t *wdt_timg_protect = (uint32_t *)TIMG_WDTWPROTECT_REG(0);
+	volatile uint32_t *wdt_timg_reg = (uint32_t *)TIMG_WDTCONFIG0_REG(0);
+
+	*wdt_rtc_protect = RTC_CNTL_WDT_WKEY_VALUE;
+	*wdt_rtc_reg &= ~RTC_CNTL_WDT_FLASHBOOT_MOD_EN;
+	*wdt_rtc_protect = 0;
+	*wdt_timg_protect = TIMG_WDT_WKEY_VALUE;
+	*wdt_timg_reg &= ~TIMG_WDT_FLASHBOOT_MOD_EN;
+	*wdt_timg_protect = 0;
+#endif
+
+#if CONFIG_BOOTLOADER_ESP_IDF
+	/* ESP-IDF 2nd stage bootloader enables RTC WDT to check on startup sequence
+	 * related issues in application. Hence disable that as we are about to start
+	 * Zephyr environment.
+	 */
+	*wdt_rtc_protect = RTC_CNTL_WDT_WKEY_VALUE;
+	*wdt_rtc_reg &= ~RTC_CNTL_WDT_EN;
+	*wdt_rtc_protect = 0;
+#endif
+
+	/* Configure the Cache MMU size for instruction and rodata in flash. */
+	extern uint32_t esp32c3_rom_cache_set_idrom_mmu_size(uint32_t irom_size, uint32_t drom_size);
+	extern int _rodata_reserved_start;
+	uint32_t rodata_reserved_start_align =
+		(uint32_t)&_rodata_reserved_start & ~(MMU_PAGE_SIZE - 1);
+	uint32_t cache_mmu_irom_size =
+		((rodata_reserved_start_align - SOC_DROM_LOW) / MMU_PAGE_SIZE) * sizeof(uint32_t);
+
+	esp32c3_rom_cache_set_idrom_mmu_size(cache_mmu_irom_size, CACHE_DROM_MMU_MAX_END - cache_mmu_irom_size);
+
+	/* set global esp32c3's INTC masking level */
+	esprv_intc_int_set_threshold(1);
+
+	/* Start Zephyr */
+	z_cstart();
+
```

Comment:
I guess this code want to avoid re-init C stack or bss areas in standard boot sequence (_PrepC() function).
However _PrepC does not only initialize the stack but also PMP, SoC depended initialization.

My concern is that can this code execute correctly even if changing implementation of _PrepC() in the future?

Reply:
Unfortunately, Xtensa (ESP32) and RISC-V is not same structure.
Xtensa does not have _PrepC() so no one can add codes. But RISC-V has _PrepC() and someone expect (Ex. PMP) to run initialization at there. If exp32c3 can guarantee to initialize all feature/modules at this _start function, it's no problem. But I think it is very uncomfortable for future esp32c3 maintainer...

What is current _PrepC() problem? If that is only BSS clear, shall we add new CONFIG for avoiding BSS clear?

Reply:
This make total sense, added the _PrepC at end of SoC specific startup code, delaying a bit the zeroing of the bss does not cause problem on the C3 special init code.

Thanks for the suggestion @katsuster